DT-6182 rental scooters

sharhio opened the pull request:
## Proposed Changes

  - Scooters added to routes and map. 
  - Scooters configured for HSL and Matka (Helsinki only)

When testing 
1. Ensure that otp is configured properly (see data-container pr)
2. UI map layer configuration uses the local otp instead of dev. 

Relies on:
https://github.com/HSLdevcom/OpenTripPlanner-data-container/pull/403
https://github.com/opentripplanner/OpenTripPlanner/pull/5909

## Pull Request Check List

  - A reasonable set of unit tests is included
  - Console does not show new warnings/errors
  - Changes are documented or they are self explanatory
  - This pull request does not have any merge conflicts
  - All existing tests pass in CI build

## Review

  - Read and verify the code changes
  - Test the functionality by running the UI locally with all popular browsers available in your platform
  - Check that the implementation matches the design, when such one is defined in a Jira issue
  - Merge the pull request

sharhio committed:
Merge remote-tracking branch 'origin/DT-6175-4' into DT-6182

sharhio committed:
Merge branch 'v3' into DT-6182

sharhio committed:
Merge branch 'v3' into DT-6182

sharhio committed:
Merge branch 'v3' into DT-6182

sharhio committed:
Merge branch 'v3' into DT-6182

sharhio committed:
Merge branch 'v3' into DT-6182

sharhio committed:
feat: dt-6182 initial scooters

sharhio committed:
feat: DT-6182 refactoring merged

sharhio committed:
feat: dt-6182 merge refatoring

sharhio committed:
merge v3

sharhio committed:
DT-6182 scooter features

sharhio committed:
DT-6182 test, style and translations fixes

sharhio committed:
DT-6182 merge, bug fix on itinerary, style fixes

sharhio committed:
DT-6182 merge v3, fix issues

sharhio committed:
fix test result after adding a mode

sharhio committed:
fix test

sharhio committed:
DT-6182 merge v3

sharhio committed:
DT-6182 fix scooter duration & scooter leave message

sharhio committed:
DT-6182 scooter return removed from walkleg, always shows

sharhio committed:
DT-6182 scooter cluster details view, show only bikestations

sharhio committed:
DT-6182 scooter lighter grey color scheme

sharhio committed:
DT-6182 search nearest by network, scooters are never default

sharhio committed:
DT-6182 remove scooters from unnecessary maplayers

sharhio committed:
merge v3

sharhio committed:
DT-6182 show only configured stations or vehicles on map + fixes

sharhio committed:
DT-6182 schema update rentalvehicle systemurl

sharhio committed:
Merge branch 'v3' into DT-6182

sharhio committed:
DT-6182 filter out invisible stations

sharhio committed:
DT-6182 ui clustering

sharhio committed:
DT-6182 do not add invisible rental stations to map data

sharhio committed:
DT-6182 sorting & settings fixed

sharhio committed:
DT-6182 cluster row selection, renaming

sharhio committed:
DT-6182 vehicle rental row tests

sharhio committed:
DT-6182 renaming

sharhio committed:
DT-6182 renaming, translations fix

sharhio committed:
DT-6182 clarification

sharhio committed:
DT-6182 style fix in relaxed scooter result

sharhio committed:
DT-6182 rentalvehicle link validation

sharhio committed:
DT-6182 removed test config

sharhio committed:
DT-6182 scooter logo color fixed

sharhio committed:
Merge branch 'v3' into DT-6182

sharhio committed:
schema fix

sharhio committed:
DT-6182 clustering dependency

sharhio committed:
DT-6182 scooter configs

sharhio committed:
merge v3

sharhio committed:
DT-6182 scooters off in layeroptions, default icon fix, taxi text removed

sharhio committed:
old test config removed

sharhio committed:
local test maplayer config removed

sharhio committed:
Merge branch 'v3' into DT-6182

sharhio committed:
fix error in merge

vesameskanen commented:
This is a mighty PR! Review will take days.

To get started, fix all new linter warnings. It is not hard - you have already added rentalVehicleShape. Just use the shapes.

vesameskanen commented on app/configurations/config.hsl.js:
```diff
@@ -127,6 +133,8 @@ export default {
       'mode-speedtram': '#007E79',
+      'mode-scooter': '#BABABA',
+      'mode-scooter-secondary': '#333333',
     },
```
secondary is probably not needed. All scooters use same icon.
Also, scooter color is already defined in base config (config.default.js)

vesameskanen commented on app/configurations/config.hsl.js:
```diff
@@ -31,6 +31,12 @@ export default {
+    },
+    REALTIME_RENTAL_VEHICLE_MAP: {
+      default: `${POI_MAP_PREFIX}/fi/realtimeRentalVehicles/`,
+    },
```
No need to define URLs again if they are same as in default config.

vesameskanen replied:
hmm my mistake, URL is actually not the same (hsl has own map path). I restored the change.

vesameskanen commented on app/server.js:
```diff
@@ -327,6 +327,9 @@ export default async function serve(req, res, next) {
+    /*    res.write(
+      '<script src="https://cdn.jsdelivr.net/gh/c-kick/mobileConsole/hnl.mobileconsole.min.js"></script>\n',
+    ); */
 
```
what is this?

sharhio replied:
An accident from debugging something...

vesameskanen commented on app/store/MapLayerStore.js:
```diff
@@ -36,12 +37,23 @@ class MapLayerStore extends Store {
+    );
+    this.mapLayers.scooter =
+      config.transportModes.scooter.hideUnlessSelectedForRouting &&
+      showRentalVehiclesOfType(
```
Please use conditional:
 config.transportModes.scooter?.hideUnlessSelectedForRouting
External UI users might not have scooter included in their conf.

vesameskanen commented on app/util/planParamUtil.js:
```diff
@@ -186,6 +202,13 @@ export function planQueryNeeded(
+        !wheelchair &&
+        settings.allowedScooterRentalNetworks > 0
+      );
+
```
Is it by design that plain scooter routes are not searched?

sharhio replied:
Yes, scooters are only supposed to be shown together with some type of public transportation.

sharhio committed:
DT-6182 use vehicleRentalSystemType for system url

sharhio committed:
DT-6182 filterByNetwork naming

sharhio committed:
filterByNetwork description

sharhio committed:
DT-6182 schema fix

sharhio committed:
DT-6182 missing icon added

sharhio committed:
DT-6182 linter warning fixes

sharhio committed:
DT-6182 transportmode shape fix

vesameskanen commented on app/component/itinerary/ItineraryPage.js:
```diff
@@ -352,6 +368,81 @@ export default function ItineraryPage(props, context) {
+      TransportMode.Scooter,
+    );
+    getPlanParams(config, match, PLANTYPE.TRANSIT, true);
+
```
I guess this line should be removed?

vesameskanen commented on app/component/itinerary/ItineraryPage.js:
```diff
@@ -352,6 +368,81 @@ export default function ItineraryPage(props, context) {
+      PLANTYPE.SCOOTERTRANSIT,
+      true, // force relaxed settings
+      //   true, // force scooter query
+    );
```
force what?

sharhio replied:
Accidental comment left from an earlier iteration.

sharhio committed:
DT-6182 config fixes, comment removed

sharhio committed:
DT-6182 removed extra method call and comment

sharhio committed:
DT-6182 scooter color fix, conditional scooter

sharhio committed:
DT-6182 support for selecting single scooter markers, cluster id fix

vesameskanen committed:
fix: restore path change caused by incorrect code review

vesameskanen commented on app/util/planParamUtil.js:
```diff
@@ -296,6 +323,10 @@ export function getPlanParams(
+    case PLANTYPE.SCOOTERTRANSIT:
+      access = ['WALK', 'SCOOTER_RENTAL'];
+      egress = access;
+      break;
```
One more change: set numItineraries to 1 if searching scooter.

sharhio replied:
The search also returns plain scooter itineraries, which are not shown on the UI. Searching for more than one itinerary gives a better chance to receive an itinerary with another mode of transport. It's not very efficient but I don't think it's possible to force the query to not return plain scooter journeys.

vesameskanen replied:
Maybe transitOnly=true option will help?

sharhio replied:
Actually, now that I checked again, the problem was that we can't force the itinerary to contain a scooter journey so it's better to search more so that at least one of the itineraries contains a scooter leg as well if it's feasible.

vesameskanen replied:
Right, that indeed makes sense.

vesameskanen commented on app/component/itinerary/ItineraryPage.js:
```diff
@@ -352,6 +368,79 @@ export default function ItineraryPage(props, context) {
+  }
+
+  async function makeRelaxedScooterQuery() {
+    if (!planQueryNeeded(config, match, PLANTYPE.TRANSIT)) {
```
I am not able to see relaxed scooter itineraries. Query is made but standard relaxed itinerary set is shown.

sharhio replied:
The relaxed scooter itineraries are only supposed to show up if no standard relaxed itineraries are found. This is a rather rare situation and I have only been able to test it by breaking the search temporarily by disabling transport modes when the standard relaxed search is made.

vesameskanen replied:
ok!

vesameskanen commented on app/component/itinerary/ItineraryPage.js:
```diff
@@ -717,6 +818,18 @@ export default function ItineraryPage(props, context) {
+      setCombinedRentalState({ plan });
+    }
+  }, [scooterState.plan, state.plan]);
+
```
There is a minor issue: when user hits earlier button, the new set is merged with standard itineraries, not with scooter-combined itineraries. Same with later button if arrive by option is set.

vesameskanen replied:
Canot duplicate. Most likely 'user error'.

vesameskanen commented on app/component/RentalVehicleContent.js:
```diff
@@ -0,0 +1,178 @@
+import PropTypes from 'prop-types';
```
Scooter page layout seems broken - e.g.  back arrow is huge and overlaps other content. I guess it should look similar to citybike page.
![image](https://github.com/HSLdevcom/digitransit-ui/assets/13776096/54179df8-9071-4dec-a4a0-165df41a6c9f)

sharhio replied:
Yeah, not sure how this was missed but I adjusted it. There are no exact design pictures for the desktop scooters, only mobile version. I also tried to make sure it closely resembles the design pictures (there was a line missing).

vesameskanen reviewed (changes requested):
PlanParamUtil should set numItineraries to 1 if searching scooters.

sharhio committed:
DT-6182 scooter details styles

vesameskanen commented on app/store/MapLayerStore.js:
```diff
@@ -36,12 +37,23 @@ class MapLayerStore extends Store {
+    );
+    this.mapLayers.scooter =
+      config.transportModes.scooter?.hideUnlessSelectedForRouting &&
+      showRentalVehiclesOfType(
```
This is a counter intuitive: I changed hideUnless... attribute to false in hsl conf and scooters disappeared. Hide name suggests  it should work the opposite way.

sharhio replied:
Ah, yes, apparently my brain works backwards, I reversed it to showIf... instead.

vesameskanen commented on app/component/itinerary/BicycleLeg.js:
```diff
@@ -244,6 +275,51 @@ export default function BicycleLeg(
+        {isScooter && !scooterSettingsOn && (
+          <div>
+            <div className={cx('itinerary-leg-action', 'scooter')}>
+              <div className="itinerary-leg-action-content">
```
I think these divs could be combined. New div is not needed just for adding one more style class.

vesameskanen commented on app/component/itinerary/CustomizeSearch.js:
```diff
@@ -127,7 +151,69 @@ class CustomizeSearch extends React.Component {
+                  <div className="transport-modes-container">
+                    {this.state.showEScooterDisclaimer && (
+                      <div className="e-scooter-disclaimer-container">
+                        <div className="e-scooter-disclaimer">
```
Can we merge disclaimer-conteiner and dislaimer, as well as respective css classes, into one? Usually the simpler the better.

vesameskanen commented on app/component/itinerary/Itinerary.js:
```diff
@@ -1019,6 +1035,21 @@ const containerComponent = createFragmentContainer(ItineraryWithBreakpoint, {
+              ios
+              web
+            }
+            vehicleRentalSystem {
```
Are these new request fields used? Request shoudl be always minimal for each component level.

sharhio replied:
Good point! Forgot to tidy this up.

sharhio committed:
remove comments

sharhio committed:
DT-6182 reverse show when routing logic

sharhio committed:
merge disclaimer divs

sharhio committed:
DT-6182 remove unnecessary fields from queries

sharhio committed:
DT-6182 scooter detail box layout fixed

vesameskanen committed:
fix: rename network list parameter

Don't call it allowedBikeRentalNetworks because it is used for scooters too.

vesameskanen committed:
fix: remove unused query fields

vesameskanen committed:
fix: remove more unused query fields

vesameskanen committed:
chore: simplify scooter itinerary querying
- rename for consistency
- scooters can use exactly similar state as other itinerary queries

vesameskanen committed:
fix: put query need checks where they belong

vesameskanen committed:
fix: scooter and transit plan merge logic

vesameskanen committed:
fix: check combinedState loading instead of scooter and transit separately

This prevents intermediate rendering before plan merging

vesameskanen committed:
Merge remote-tracking branch 'origin/v3' into DT-6182

vesameskanen committed:
fix: triplink props

vesameskanen committed:
fix: remove unused rentalVehicle fields from itineraryLine

vesameskanen committed:
fix: remove OTP1 modes

vesameskanen committed:
fix: refactor badly outdated transport mode handling

UI contained lots of complex code for adding citybike and scooter to transport modes array.
In the end, those were just filtered away before making a request. Most of the old mess
is now cleaned away. There is still some strange code related to availableForSelection config switch;
it should probably be erased as well.

vesameskanen committed:
chore: update modeUtil unit tests

vesameskanen committed:
fix: remove useless modeToOtp util func

It does nothing.

vesameskanen committed:
chore: refactor MapLayersDialogContent.js

vesameskanen committed:
fix: no need for conditional config access

We could add ? to every object reference but would it make the code better?

vesameskanen committed:
fix: unit tests whic did not pass config

vesameskanen committed:
fix: do not suggest citybike if wheelchair is enabled

Wheelchair option disables biking and scooters. Do the same with citybikes.

vesameskanen committed:
fix: do not suggest citybikes if wheelchair option is set

vesameskanen committed:
chore: rename

vesameskanen committed:
fix: catch hsl banner fetch errors

vesameskanen committed:
fix: check correct plan when considering relaxed plans

vesameskanen committed:
fix: no need to expose children's prop in container

vesameskanen committed:
fix: no need to pass itineray page state back as parameter from a child callback

vesameskanen committed:
fix: pass openSettings func, not toggle, to scooter legs

The button in scooter leg says 'Open settings'. Toggle will actually
close the panel if it is already open.

vesameskanen committed:
fix: change toggleSettings to openSettings in unit tests

vesameskanen committed:
fix: rename network types

vesameskanen committed:
fix: remove local test URLs from config

vesameskanen committed:
fix: unit tests

vesameskanen commented:
I added some minor fixes and refactored the code. There might be still a need for renaming more citybike things to a  generic form.
